### PR TITLE
Allow user defined changes() and suffix as query

### PR DESCRIPTION
**Pull request by birkir**
For example sorting (after changes):
```javascript
class Turtles extends Component {

  observe (props, state) {

    const query = r.table('turtles')
    .sortBy({
      index: 'name'
    }).changes({
      includeInitial: true,
      includeState: true
    })
    .skip(10)
    .limit(5);

    return {
      turtles: new QueryRequest({
        query,
        changes: true
      }),
    };
  }
}
```

We could do a `query.toString().match(/\.changes\(.*?\)/)` to not break the api, or allow optional callback function to apply the `changes()` to the query.

### CODE REVIEW

**Comment by mikemintz** [comment]
@birkir is this only for working around the isomorphic screen flicker in https://github.com/mikemintz/react-rethinkdb/issues/29, or are there other use cases? I'd prefer to solve that issue properly by having changefeed resume support in rethinkdb, to avoid the race condition.

**Comment by birkir** [comment]
Hi @mikemintz, this is a seperate issue.

First of all, the user should be able to set `changefeedQueueSize` and `squash` property to the changes method for optimisation purposes.

The other use-case is where transformations are applied **after** the changes() method. You can add a `filter()`, `limit()`, `skip()` and more methods that are necessary to be placed **after** the changes() method to work.

https://rethinkdb.com/docs/changefeeds/javascript/#changefeeds-with-filtering-and-aggregation-queries

**Comment by mikemintz** [comment]
Ok yeah, I see that can be useful now.

So currently, we look at `queryRequest.changes` to determine two things: 1) whether to use `_runStaticQuery` or `_runChangeQuery`, and 2) how to modify the query

I think for 1, we can probably combine those into one method, and branch on `cursor.constructor.name` to determine whether it's a changefeed query or not.

For 2, I'd like to keep the query transformation there, both for backwards compatibility, and to reduce boilerplate for users that have simple queries.

With those changes, you could do your use case by explicitly specifying `.changes(...)` in the observe method, and not specifying `changes: true`. All other use cases would work the same as before.

Does this seem like a reasonable solution?

For determining the feed type, take a look at:

* https://github.com/rethinkdb/rethinkdb/blob/next/drivers/javascript/net.coffee#L435-L445
* https://rethinkdb.com/docs/writing-drivers/ (under Response Notes)
* https://github.com/mikemintz/react-rethinkdb/blob/master/src/QueryState.js#L78

**Comment by birkir** [comment]
Hi @mikemintz, that's a very good and reasonable solution and even simplifies the react-rthinkdb architecture (at least from my perspective). As I'm against breaking api's I am 100% with you on this.

File: src/QueryState.js [hunk 1]
```diff
@@ -54,33 +50,19 @@ export class QueryState {
...
     this.closeHandlers.push(() => promise.then(x => isCursor(x) && x.close()));
     promise.then(cursor => {
+      const isFeed = !!cursor.constructor.name.match(/Feed$/);
       const isPointFeed = cursor.constructor.name === 'AtomFeed';
-      this.value = isPointFeed ? undefined : [];
-      cursor.each((error, row) => {
-        if (error) {
-          this._addError(error);
-        } else {
+      if (isFeed) {
```
Comment (mikemintz) [inline src/QueryState.js]: Can you move this up one line so `const isPointFeed = ...` is only run for the `if (isFeed)` then clause?

File: src/QueryState.js [hunk 2]
```diff
@@ -54,33 +50,19 @@ export class QueryState {
...
+        this.value = isPointFeed ? undefined : [];
+        cursor.each((error, row) => {
+          if (error) {
+            return this._addError(error);
```
Comment (mikemintz) [inline src/QueryState.js]: Any reason not to keep the original logic of putting the code below in the `else` block and not having a `return` here?

I'm guessing it's probably the same behavior, but safer to change less, and I think using `return` for a value that's always undefined is a bit misleading.

Comment (birkir) [inline src/QueryState.js]: Sure, just a thing I often do automatically when nested if statements get to deep (4 levels in this case). Safer is better.

File: src/QueryState.js (hunk 1 above)
Comment (birkir) [inline src/QueryState.js]: :+1:

**Comment by mikemintz** [comment]
Looking good, this will be great to merge for flexibility! Could you rebase squash so it's just one commit (named like the "Move static and change query methods to a hybrid one" one), and then I'll go ahead and merge it in?

**Comment by birkir** [comment]
Good to go!

**Pull request merged**